backend/oai.py

```diff
@@ -157,33 +204,47 @@ def process_record(self, header, metadata, format):
             except ValueError:
                 logger.exception("Ignoring invalid paper with header %s" % header.identifier())
 
-    def process_records(self, listRecords, format):
+    def process_records(self, listRecords, metadata_format, max_lookahead=1000):
         """
-        Save as :class:`Paper` all the records contained in this list
+        Save as :class:`Paper` all the records contained in this list.
+        Records are represented as pairs of OaiHeader and OaiRecord, as returned
+        by pyoai's ListRecords
         """
         # check that we have at least one translator, otherwise
         # it's not really worth trying…
         if not self.translators:
             raise ValueError("No OAI translators have been set up: " +
                              "We cannot save any record.")
 
-        last_report = datetime.now()
-        processed_since_report = 0
-
-        with ParallelGenerator(listRecords, max_lookahead=1000) as g:
-            for record in g:
-                header = record[0]
-                metadata = record[1]._map
-
-                self.process_record(header, metadata, format)
-
-                # rate reporting
-                processed_since_report += 1
-                if processed_since_report >= 1000:
-                    td = datetime.now() - last_report
-                    rate = 'infty'
-                    if td.seconds:
-                        rate = str(processed_since_report / td.seconds)
-                    logger.info("current rate: %s records/s" % rate)
-                    processed_since_report = 0
-                    last_report = datetime.now()
+        with ParallelGenerator(listRecords, max_lookahead=max_lookahead) as g:
+            for record_group in group_by_batches(with_speed_report(g, name='OAI papers')):
+                oai_ids_to_last_updated = {
+                    record[0].identifier():record[0].datestamp() for record in record_group
+                }
+
+                # Fetch the last modified date of all these records in the DB
+                last_modified_in_db = { id : last_modified
+                    for id, last_modified in
+                    OaiRecord.objects.filter(identifier__in=oai_ids_to_last_updated.keys()).values_list('identifier', 'last_update')
+                }
+
+                # Deduce the list of papers to update
+                ids_to_update = set()
+                for id, last_updated_in_base in oai_ids_to_last_updated.items():
+                    if id not in last_modified_in_db or last_modified_in_db[id].date() <= last_updated_in_base.date():
+                        ids_to_update.add(id)
```

Comment:
Is this `for` loop large? Might be (a bit) more efficient to do it with a generator on the fly:

```python
ids_to_update = set(
    ident
    for ident, last_updated_in_base in oai_ids_to_last_updated.items() 
    if ident not in last_modified_in_db or last_modified_in_db[ident].date() <= last_updated_in_base.date()
)
```